### PR TITLE
Add ability to clean the DB

### LINKED ISSUE #583: Consider offering a way to reset DB at Hibernate ORM level for testing
reset would rerun the DDL statements and import.sql file.

This is useful if you want a test to be fully isolated from another data wise.

### PR DESCRIPTION
**Pull request by stuartwdouglas**
- In tests this is done via @ResetDatabase
- In dev mode this is done from the dev console

Fixes #583

### CODE REVIEW

**Comment by stuartwdouglas** [comment]
I realized after I had this working that it should also integrate with Flyway/Liquibase, so just a draft for now.

**Comment by yrodiere** [comment]
This feature looks nice! I have two questions:

* Shouldn't this reset the database *before* the test? If I annotate a given test method, I'm mostly interested in what happens during this particular test...
* Could this be adapted to reset persisted data that is not strictly speaking in a database, e.g. distributed caches, MongoDB, Elasticsearch, ...? And if so, shouldn't the name be more generic, such as `@ResetPersistence`? `@ResetData`, maybe?

**Comment by famod** [comment]
Nice!

Are you aware of #14240?
I wonder if that `DatabaseSchemaProvider` approach could be extended/used to allow a DB dump/restore with DB-specific tools like outlined in https://github.com/quarkusio/quarkus/issues/14240#issuecomment-758842298
The scenario would be that you have `quarkus-liquibase` in the project and if the DB is blank in the beginning, Liquibase shall be executed, but everything after that should use a much quicker dump/restore mechanism...

File: extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AgroalRecorder.java [hunk 1]
```diff
@@ -29,4 +35,20 @@ public AgroalDataSource get() {
...
+        return new DevConsolePostHandler() {
+            @Override
+            protected void handlePost(RoutingContext event, MultiMap form) throws Exception {
+                String name = form.get("name");
+                ServiceLoader<DatabaseSchemaProvider> dbs = ServiceLoader.load(DatabaseSchemaProvider.class,
+                        Thread.currentThread().getContextClassLoader());
+                for (DatabaseSchemaProvider i : dbs) {
+                    i.resetDatabase(name);
+                }
+                flashMessage(event, "Action invoked");
```
Comment (FroMage) [inline extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AgroalRecorder.java:50]:
```suggestion
                flashMessage(event, "Database "+name+" reset");
```

File: extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java [hunk 2]
```diff
@@ -333,4 +347,36 @@ HealthBuildItem addHealthCheck(DataSourcesBuildTimeConfig dataSourcesBuildTimeCo
...
+        return new DevConsoleRouteBuildItem("clean", "POST", recorder.devConsoleCleanDatabaseHandler());
+    }
+
+    @BuildStep(onlyIf = IsTest.class)
+    void cleanDatabaseSupport(BuildProducer<GeneratedBeanBuildItem> generatedBeanBuildItemBuildProducer,
+            BuildProducer<AdditionalBeanBuildItem> additionalBeans) {
+        //generate the annotated interceptor with gizmo
+        //all the logic is in the parent, but we don't have access to the
+        //binding annotation here
+        try (ClassCreator c = ClassCreator.builder()
```
Comment (FroMage) [inline extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java:371]: The shit we do to deal with modularity :(

File: extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/HibernateSchemaRecreateDevConsoleTestCase.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+                    .addClasses(MyEntity.class, MyEntityTestResource.class)
+                    .addAsResource("application.properties")
+                    .addAsResource("import.sql"));
+
+    @Test
+    public void testCleanDatabase() {
+        RestAssured.when().get("/my-entity/count").then().body(is("2"));
+        RestAssured.when().get("/my-entity/add").then().body(is("MyEntity:added"));
+        RestAssured.when().get("/my-entity/count").then().body(is("3"));
+        RestAssured.with()
```
Comment (FroMage) [inline extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/HibernateSchemaRecreateDevConsoleTestCase.java:26]: I never realised that tests could access dev mode actions.

**Comment by FroMage** [comment]
That PR is awesome!

I do agree with @yrodiere that I was expecting the DB to be cleaned _before_ my tests. And also that a similar mechanism could be added for other things like caches, but I guess that this can wait until later.

**Comment by Sanne** [comment]
nice!

We'll have to be careful to allow us to be smart under the hood though, as some DBs are painfully slow to deal with schema changes; just yesterday on the ORM zulip channels we were talking about a (TBD) feature which could drop the data only rather than removing and re-creating the schema.

**Comment by stuartwdouglas** [comment]
I would expect tests to leave the DB in a consistent state by default, so there should be no need to clean up beforehand. This annotation is basically saying 'I am going to leave the DB in an inconsistent state, I want you to clean it up for me'

**Comment by famod** [comment]
Sometimes, it can be handy to check in the DB what the test has left behind without having to debug it.
Temporarily removing the annotation leaves you with a dirty state for the next run, so that's not an option.
Maybe an annotation parameter could control when the cleanup happens?

**Comment by yrodiere** [comment]
> I would expect tests to leave the DB in a consistent state by default, so there should be no need to clean up beforehand. This annotation is basically saying 'I am going to leave the DB in an inconsistent state, I want you to clean it up for me'

If we assume that this annotation is only useful within the scope of a single test class, because every test class always resets the database(s) before executing, then yes I agree.

Otherwise, I'm afraid that's wishful thinking: people will end up in a situation where the DB is in an inconsistent state eventually. Be it because the application crashed during the last test execution (OOM, stackoverflow), because you stopped the JVM while debugging, ...

Also... what do you think about changing the name to allow it to cover non-DB cleanups in the future? `@ResetData`, `@ResetPersistence`, whatever works?

**Comment by FroMage** [comment]
> I would expect tests to leave the DB in a consistent state by default, so there should be no need to clean up beforehand. This annotation is basically saying 'I am going to leave the DB in an inconsistent state, I want you to clean it up for me'

That's optimistic at best ;)

In practice I always have some tests that don't give a damn about the initial state of the DB, so they can all run without any cleanup. And then some really care and need a clean DB before them.

It's true that I could fix both tests by cleaning up after them. BTW, is there a way to tell a test class to cleanup after all tests versus after each test?

**Comment by stuartwdouglas** [comment]
Because this is an interceptor I can't really do 'after the class', only 'after each method', it would need a different implementation if we were going to support the 'after class' approach.

I have also been thinking about native tests, and what we can do to support it. It's a lot more complex, as you don't really want a way for the test to trigger a reset of the DB (as this could be a huge security issue, if there is a way to tell a production instance to clean the production DB).

**Comment by stuartwdouglas** [comment]
I have opened https://github.com/quarkusio/quarkus/pull/19312 for just the Dev UI part

**Comment by stuartwdouglas** [comment]
Closing this, as support for added to the dev console

**Pull request closed without merging**